web/scripts/checkout/checkoutComponent.js

```diff
@@ -0,0 +1,142 @@
+              </div>
+              <br>
+              <button class="btn btn-primary">Submit Payment</button>
+            </form>
+          </div>
+          <div class="emptyCart" v-else><h2>Your cart is empty</h2></div>
+        </div><!-- /component -->`,
+      data: () => {
+        return state
+      },
+      mounted: () => {
+        if (state.products.length > 0) {
+          // Create stripe card input.
+          card = state.stripe.elements().create('card')
+          card.mount('#card-element')
+          // Add event listener for input errors in real time.
+          const displayError = document.getElementById('card-errors')
+          message = displayError
+          card.addEventListener('change', function (event) {
+            if (event.error) {
+              displayError.textContent = event.error.message
+            } else {
+              displayError.textContent = ''
+            }
+          })
+          // Add event listener to handle payment processing with stripe.
+          const form = document.getElementById('payment-form')
+          form.addEventListener('submit', function (event) {
+            event.preventDefault()
+            if (storage.get('user')) {
+              createToken()
+            } else {
+              displayError.textContent = 'Please login to complete your purchase'
+            }
+          })
+        }
+      }
+    })
+
+    function createToken () {
+      state.stripe.createToken(card).then(function (result) {
+        if (result.error) {
+          // Inform the customer that there was an error.
+          var errorElement = document.getElementById('card-errors')
+          errorElement.textContent = result.error.message
+        } else {
+          // Send the token to the server.
+          const checkout = new Checkout({
+            email: storage.get('user').email,
+            tokenID: result.token.id,
+            total: state.subtotal,
+            products: state.products
+          })
+          checkoutRepo.charge(checkout, (err, res) => {
+            if (err) {
+              console.log(err)
+              console.log('Charge failed')
+              return
+            }
+
+            // Empty the shopping cart.
+            while (shoppingCart.getItems().length !== 0) {
+              shoppingCart.removeItem(state.products[0])
+            }
+
+            // Reload purchase history
+            usersRepo.getHistory(storage.get('user').email, (err, res) => {
+              if (err) {
+                console.log('Check connection')
+                return
+              }
+
+              storage.set('purchaseHistory', res)
+            })
+            router.navigate('/checkout-success')
+          })
+          message.textContent = 'Please wait while we process your order'
+          // TODO: Are we allowing user to download a fake e-book?
+          // If so, this can be done from user profile.
+          // Should we add tax?
+        }
+      })
```

Comment:
This function has a lot of lines in it, which makes it harder to understand, and to test. It's also non-deterministic because of things like `state.stripe.createToken(card)`, and `new Checkout(...)`. Here is an example that replaces `createToken` with deterministic functions, which can be exported and tested:

```JavaScript
const makeCallbackPromise = (resolve, reject) => (err, res) => {
  if (err) {
    return reject(err)
  }

  return resolve(res)
}

const showLoadingState = () => {
  message.textContent = 'Please wait while we process your order'
}

const makeToken = (state, card) => {
  return state.stripe.createToken(card)
}

const checkForErrors = (result) => {
  if (result.error) {
    throw new Error(result.error.message)
  }

  return result
}

const makeCheckout = (email, state) => (result) => {
  return new Checkout({
    email: email,
    tokenID: result.token.id,
    total: state.subtotal,
    products: state.products
  })
}

const executeTransaction = (checkoutRepo) => (checkout) => {
  return new Promise((resolve, reject) => {
    checkoutRepo.charge(checkout, makeCallbackPromise(resolve, reject))
  })
}

const clearShoppingCart = (shoppingCart, state) => () => {
  while (shoppingCart.getItems().length !== 0) {
    shoppingCart.removeItem(state.products[0])
  }
}

const getPurchaseHistory = (email, usersRepo) => () => {
  return new Promise((resolve, reject) => {
    usersRepo.getHistory(email, makeCallbackPromise(resolve, reject))
  })
}

const storePurchaseHistory = (storage) => (history) => {
  storage.set('purchaseHistory', history)

  return history
}

const navigateToSuccess = (router) => () => {
  router.navigate('/checkout-success')
}

const createToken = (options) => {
  // by allowing the caller to override the options, this function
  // is deterministic when they are presented, and non-deterministic
  // when they are not: both testable, and self executing
  options = Object.assign({
    email: storage.get('user').email,
    card: card,
    state: state,
    shoppingCart: shoppingCart,
    storage: storage,
    router: router,
    checkoutRepo: checkoutRepo,
    usersRepo: usersRepo
  }, options)

  return Promise.resolve(showLoadingState())
    .then(() => makeToken(options.state, options.card))
    .then((result) => checkForErrors(result))
    .then((result) => makeCheckout(options.email, options.state)(result))
    .then((checkout) => executeTransaction(options.checkoutRepo)(checkout))
    .then(() => clearShoppingCart(options.shoppingCart, options.state)())
    .then(() => getPurchaseHistory(options.email, options.usersRepo)())
    .then((history) => storePurchaseHistory(options.storage)(history))
    .then(() => navigateToSuccess(router)())
    .catch((err) => {
      // NOTE: it would be better to import a card errors component
      var errorElement = document.getElementById('card-errors')
      errorElement.textContent = err.message
    })
}
```